### PR TITLE
Allow deprecation of input values

### PR DESCRIPTION
**Pull request by smitt04**
Started the @deprecated proposal in reference to #197

Please let me know if there is anything i missed or you would like changed.  This was just a quick update to get the ball rolling again.

@IvanGoncharov Are you still interested in reviewing this?

### CODE REVIEW

File: spec/Section 3 -- Type System.md [hunk 1]
```diff
@@ -1778,3 +1778,15 @@ type ExampleType {
   oldField: String @deprecated(reason: "Use `newField`.")
 }
 ```
+
```
Comment (IvanGoncharov) [inline spec/Section 3 -- Type System.md]: I think you also need to expand above text:
> The @deprecated directive is used within the type system definition language to indicate deprecated portions of a GraphQL service’s schema, such as deprecated fields on a type or deprecated enum values.

File: spec/Section 4 -- Introspection.md [hunk 2]
```diff
@@ -332,6 +334,8 @@ Fields
 * `name` must return a String.
 * `description` may return a String or {null}.
 * `inputFields`: a list of `InputValue`.
+  * Accepts the argument `includeDeprecated` which defaults to {false}. If
+    {true}, deprecated enum values are also returned.
```
Comment (IvanGoncharov) [inline spec/Section 4 -- Introspection.md]:
```suggestion
    {true}, deprecated fields are also returned.
```

File: spec/Section 4 -- Introspection.md [hunk 3]
```diff
@@ -392,6 +396,9 @@ Fields
 * `defaultValue` may return a String encoding (using the GraphQL language) of the
   default value used by this input value in the condition a value is not
   provided at runtime. If this input value has no default value, returns {null}.
+* `isDeprecated` returns {true} if this field or argument should no longer be used,
+  otherwise {false}.
+* `deprecationReason` optionally provides a reason why this field or argument is deprecated.
```
Comment (IvanGoncharov) [inline spec/Section 4 -- Introspection.md]:
```suggestion
* `deprecationReason` optionally provides a reason why this input field or argument is deprecated.
```

**Comment by IvanGoncharov** [comment]
@smitt04 Looks good so far, I left a few review feedbacks.
Next step would be to work on PR for `graphql-js`: https://github.com/graphql/graphql-js
Just create a draft PR and I will review it.

Also, we need to discuss it on WG, here is the next one:
https://github.com/graphql/graphql-wg/blob/master/agendas/2018-11-08.md
You can add yourself to attendees and a new item to agenda like this:
https://github.com/graphql/graphql-wg/pull/103/files

**Comment by smitt04** [comment]
@IvanGoncharov I have update the PR with suggested fixes. I have also submitted a PR to graphql-wg.

I can start looking into graphql-js, although I haven't worked in javascript graphql for a couple years.  I am mostly working with golang graphql as of late.

**Comment by IvanGoncharov** [comment]
> I can start looking into graphql-js, although I haven't worked in javascript graphql for a couple years. I am mostly working with golang graphql as of late.

NP, you can just open an issue on `graphql-js` and I will try to implement it before WG.

**Comment by tomasaschan** [comment]
I just want to cheer you on and add my hopes that this will be merged and added soon. 🙌 

It's quite a wonderful feeling to realize you're missing a feature, and when you start looking for discussions about it you find this (on graphql/graphql-js#1560):

image

**Comment by Neitsch** [comment]
Should we only allow the deprecated directive on nullable arguments? This can not be expressed in the directive definition, but logically that might make sense.

**Comment by Anufant** [comment]
@IvanGoncharov 
Hi Ivan! 
Could you tell me please, when this PR will be merged?
I want to deprecate some input arguments in my GQL schema very much :)

**Comment by ydorea** [comment]
I'm also really looking forward to this feature!

**Comment by IvanGoncharov** [comment]
@Anufant We discussed it on the last WG and everyone agreed that this issue should be fast-tracked so it definitely going into next spec release. Not sure about the exact date since there are a few things we need to discuss and implement in `graphql-js` before merging this.
You can read about the necessary RFC stages here: 
https://github.com/facebook/graphql/blob/master/CONTRIBUTING.md#rfc-contribution-champions

**Comment by tjkirch** [comment]
Awesome, glad this is moving!  Does that mean the "Next Stage?" tag can be updated to a later stage?  :)

**Comment by dice14u** [comment]
What is the next step here, it seems it has a champion with smitt04 having written both the propsal here and a (now stale) implementation in graphql-js "discussed it on the last WG and everyone agreed .. definitely going into next spec release" sounds like acceptance by WG so stage 3.  Is there something a community member can do to get this merged in on the fast tracked path ?

**Comment by KenanSulayman** [comment]
Revisiting. Can this be merged soon?

**Comment by xuorig** [comment]
👍 very much looking towards this. We've been "faking" argument deprecation (through argument descriptions) and its been working well.

What does it mean for a non-null argument/input value to be deprecated? I'm not seeing this discussed anywhere. At the very least implementations should probably not allow it without making the argument optional first, but I'm not sure what it means in terms of definition only. At GitHub our deprecation DSL warns you to make that field optional first.

  - Should a GraphQL server demote a required argument to nullable in practice when it is deprecated?
  - Is a required input value that is deprecated a valid state for a schema to be in?

**Comment by axelson** [comment]
Is it possible to get a status update on this? (was going to ask why "CLA Signed" was removed, but it looks like that entire label was removed so that makes sense)

**Comment by IvanGoncharov** [comment]
Hi @smitt04 
Sorry for the silence.
@andimarek brought this topic on the last WG and consensus was that there are no blockers We just need to update this PR and `graphql-js` PR and can move it to Stage2 on the next WG.
Do you still have time to work on this feature and present it on the next WG?
If you want I can help with `graphql-js` PR and rebase it on top current master?

**Comment by smitt04** [comment]
@IvanGoncharov I updated this PR and it would be great if you can update `graphql-js`

**Comment by modosc** [comment]
@IvanGoncharov @smitt04 if you need a hand let me know, i'm happy to help out to get this moving forward.

**Comment by sonnenhaft** [comment]
pls merge

**Comment by leebyron** [comment]
This needs a rebase and some cleanup before it can be considered Stage 2

**Comment by lifeiscontent** [comment]
I can't wait for this 🤩

**Comment by MidnightDesign** [comment]
> I can't wait for this 🤩

Please use reactions instead of spamming peolpe who are watching this issue.

**Comment by leebyron** [comment]
Capturing discussion notes from WG meeting:

There is one remaining pitfall from this change, and that's deprecating a **required** argument or input field. There are two problems with this:

1. Deprecations are intended to be actionable: a client should be able to edit queries to contain no deprecated elements. It would not be possible to remove a deprecated required argument.

2. Introspection does not include deprecated portions of the schema by default, you must pass `(includeDeprecated: true)` to the relevant potion. This could result in a scenario where a client omits the deprecated portion of a schema from introspection and therefore does not know about an required argument. What appears valid during client development is rejected as invalid by the service, a surprising outcome. *Note that this is not an abstract concern since no current tool asks for deprecated arguments or input fields and so all current tools could suffer this scenario.*

The recommended addition to finalize this spec proposal (and linked PR) is to include a schema validation rule whereby any deprecated argument or input field is not also required. There is an open question whether this should be a "must" or a "should" rule, which @eapache will help investigate - either way the reference implementation GraphQL.js will include this schema validation rule.

**Comment by eapache** [comment]
Apologies that it took me much longer than expected to get back to this. I've discussed this with my team internally and here's the current state of my thoughts:

> 1. Deprecations are intended to be actionable.

I don't think this is really a problem. Even setting aside our local schema versioning system (which was the motivating counter-example), not all deprecations are going to be actionable "inside the schema"; GraphQL should support deprecating required arguments or inputs where the action to be taken by the client is more complicated than just "use another field".

In our particular case, the action for any deprecation (even for output fields) always includes "upgrade your API version" (effectively switching to a different snapshot of the schema) alongside any necessary field/argument changes.

> 2. Introspection does not include deprecated portions of the schema by default.

This is a much bigger problem. It wouldn't surprise me if some developers use introspection dynamically in production, which would mean that simply deprecating a required input becomes a live breaking change for those clients.

---

My answers to these two problems are notably in conflict. For the purpose of supporting deprecations with actions "outside the schema", I do believe GraphQL should permit deprecating required inputs. However, this is a significant problem for tooling and introspection.

One option I think worth considering is simply changing the rules for introspection: if introspection always returns required parts of the schema by default (even when they're deprecated and `includeDeprecated: false`) then that problem vanishes. It does complicate deprecation a little bit, but in return you get increased expressivity (required arguments can be marked deprecated) and the underlying heuristic (introspection should always tell you the things you absolutely need to know) seems sound.

If that isn't acceptable, then I think making this a "MUST" validation is ok. There will be some cases on some schemas where `@deprecated` won't be usable, but it will still cover 99% of cases without any extra complexity.

**Comment by eapache** [comment]
A brief summary of the discussion at the working group meeting: there was resistance to making changes or adding inconsistencies to introspection. In general, the GraphQL spec is pushing a very specific model of how schemas should evolve, and doesn't strongly feel a need to support alternative models. There was a proposal to use `SHOULD` language anyway, but given the introspection problem we probably wouldn't use it, so the conclusion was to just use `MUST` as originally proposed.

**Comment by vlad-accedo** [comment]
wow

**Comment by smyrick** [comment]
FYI for those tracking the issue, it has not officially merged into the spec yet, it has just moved to the next stage. See the actual changes here: https://github.com/graphql/graphql-spec/pull/805

**Pull request merged**